### ISSUE TITLE
[Blueprint] Homebrew casks report auto_updating casks

### ISSUE THREAD

**Issue by DocEmmetBrown** [osquery/osquery#8476]
# Blueprint

Related to https://github.com/osquery/osquery/issues/8468:
TL;DR: Some Homebrew Casks have the `auto_update` flag set to true, which according to the [docs](https://docs.brew.sh/Cask-Cookbook) 
> true. Asserts that the cask artifacts auto-update. Use if Check for Updates… or similar is present in an app menu, but not if it only opens a webpage and does not do the download and installation for you.

This means that fetching the cask version from the path (as in `/opt/homebrew/Caskroom/arduino/1.8.19` for example) is wrong, since the "hardcoded" version in the path is the version that was installed, not necessarily the current one. (more details in linked issue https://github.com/osquery/osquery/issues/8468)

## Proposals

I would propose to change the way homebrew cask version is reported, to reflect this auto_update feature. 
But as I see it, there are multiple options, and I'm not sure which one would be the best: 

1: No change, continue reporting the version according to the path.

2: Fetch the "actual" version of the installed app, in case of a symlink. This raises the question of "what" the actual version is: should we call some of the https://github.com/osquery/osquery/blob/master/osquery/tables/system/darwin/apps.mm functions to fetch the proper version ? 

3: Check if the installed cask is actually a symlink to `/Applications` (or some known list of directories containing applications), and skip it from the `homebrew_package` table, since it's already checked by the apps table.

4: Add a specific column in the `homebrew_package` table for casks that are a link to an app (so it can be filtered out by Fleet / Kolide... for compliance reporting). 

5: Use the [homebrew metadata](https://github.com/Homebrew/brew/blob/41be66fc2a3facc5a575f98ffa2a17323ef30c22/Library/Homebrew/cask/cask.rb#L175) to fetch the `auto_update` flag, and  
5.1: add a new column with a boolean for whether or not a cask auto_updates
5.2: replace the version by `auto_update`.

## Opinions

My personal opinion on these options, with that caveat that I have minimal experience with the osquery codebase, and barely any with c++

1: not ideal

2: What would be the criteria if we have multiple files in the directory, one of them being a symlink `.app` file, but with other files present ? 🤔 

3: Same issue as with the previous option

4: Breaking change to the `homebrew_packages` table

5: my preferred option, easier to implement. 
5.1: breaking change, but we keep the (useful ?) information of what version was installed
5.2: We basically "ignore" the casks that are auto-updating.

## Closing thoughts
When using `osquery` and `fleet`, for compliance purposes, the fact that the "wrong" version is reported shows a wrong idea of the current state of the machine ; One can be completely up-to-date, and still have a lot of casks wrongfully reported as outdated. It would be great to have a better solution. 

I'm very much happy to discuss all the options, and I'm also probably missing other options, since I'm not familiar with osquery. We can also take it to Slack if needed, for more synchronous interactions. 

I have a quick and dirty PoC code to implement option 5.2 that I could push as a PR if it's any valuable. 

Thanks for all the hard work on this great project!

**Comment by directionless** [comment]
Nosing around, some examples:
```
osquery> select * from homebrew_packages where name = 'secretive';
+-----------+----------------------------------+---------+------+
| name      | path                             | version | type |
+-----------+----------------------------------+---------+------+
| secretive | /opt/homebrew/Caskroom/secretive | 2.4.1   | cask |
+-----------+----------------------------------+---------+------+

osquery> select CONCAT_WS('/', path, version) as path from homebrew_packages where name = 'secretive';
+----------------------------------------+
| path                                   |
+----------------------------------------+
| /opt/homebrew/Caskroom/secretive/2.4.1 |
+----------------------------------------+

osquery> select * from file where directory = (select CONCAT_WS('/', path, version) as path from homebrew_packages where name = 'secretive');
+------------------------------------------------------+----------------------------------------+---------------+-----------+-----+-----+------+--------+------+------------+------------+------------+------------+------------+------------+---------+-----------+-----------+
| path                                                 | directory                              | filename      | inode     | uid | gid | mode | device | size | block_size | atime      | mtime      | ctime      | btime      | hard_links | symlink | type      | bsd_flags |
+------------------------------------------------------+----------------------------------------+---------------+-----------+-----+-----+------+--------+------+------------+------------+------------+------------+------------+------------+---------+-----------+-----------+
| /opt/homebrew/Caskroom/secretive/2.4.1/Secretive.app | /opt/homebrew/Caskroom/secretive/2.4.1 | Secretive.app | 171799502 | 501 | 80  | 0755 | 0      | 96   | 4096       | 1706149013 | 1706149013 | 1732161171 | 1706149013 | 3          | 1       | directory |           |
+------------------------------------------------------+----------------------------------------+---------------+-----------+-----+-----+------+--------+------+------------+------------+------------+------------+------------+------------+---------+-----------+-----------+

osquery> select * from file where path = '/Applications/Secretive.app';
+-----------------------------+---------------+---------------+-----------+-----+-----+------+--------+------+------------+------------+------------+------------+------------+------------+---------+-----------+-----------+
| path                        | directory     | filename      | inode     | uid | gid | mode | device | size | block_size | atime      | mtime      | ctime      | btime      | hard_links | symlink | type      | bsd_flags |
+-----------------------------+---------------+---------------+-----------+-----+-----+------+--------+------+------------+------------+------------+------------+------------+------------+---------+-----------+-----------+
| /Applications/Secretive.app | /Applications | Secretive.app | 171799502 | 501 | 80  | 0755 | 0      | 96   | 4096       | 1706149013 | 1706149013 | 1732161171 | 1706149013 | 3          | 0       | directory |           |
+-----------------------------+---------------+---------------+-----------+-----+-----+------+--------+------+------------+------------+------------+------------+------------+------------+---------+-----------+-----------+
```

So we can see it's installed, and a symlink, but osquery isn't exposing the symlink target

**Comment by directionless** [comment]
We talked about this in office hours today. And there were a couple of perspectives to navigate...
1. There's a lot of interest in making this do what people expect
2. There is also interest in preserving the information that homebrew has -- if it was installed as version X, maybe that matters, to someone. 
3. There's general interest in exposing more information

And we had a couple of take aways:

1. Exposing `auto_update` as a column seems reasonable

2. We _should_ but do not, currently expose enough information to do this via joining the app table. We should remedy this by exposing a symlink's target in the `file` table. 

3. There's value in presenting what a user expects -- to that end, perhaps we should add a column to `homebrew_packages` for the upgrade version. Unclear what to name it. But my bias is a new column.

4. It's not going to be possible to resolve all caskroom versions. But, if there is a single app bundle, it seems reasonable to try.

**Comment by michael-myers** [comment]
Yea, apparently Homebrew has no information about the actual current version of a cask installed from a formula with `auto_updates = true`. 

I have [one such Homebrew package](https://github.com/Homebrew/homebrew-cask/blob/HEAD/Casks/a/alt-tab.rb), `alt-tab`, which I apparently installed at version 6.64.0, but no matter how recently I `brew update; brew upgrade`, is still listed as 6.64.0 because it has `auto_updates`. Homebrew will list the latest _available_ version (misleadingly, since it is not pulled from the locally installed version) via `brew info`, but the path in `/usr/local/Caskroom` is still `6.64.0`. In the app itself, I have self-updated to 7.6.0, but Homebrew doesn't know this.

```shell
brew info alt-tab
==> alt-tab: 7.6.0 (auto_updates)
https://alt-tab-macos.netlify.app/
Installed
/usr/local/Caskroom/alt-tab/6.64.0 (31.6MB)
From: https://github.com/Homebrew/homebrew-cask/blob/HEAD/Casks/a/alt-tab.rb
==> Name
AltTab
==> Description
Enable Windows-like alt-tab
==> Artifacts
AltTab.app (App)
==> Analytics
install: 6,256 (30 days), 16,281 (90 days), 60,275 (365 days)
```

The information that this cask is self-updated is present locally in the metadata files, as mentioned:

```shell
grep 'auto_updates' /usr/local/Caskroom/alt-tab/.metadata/6.64.0/20231025061257.815/Casks/alt-tab.json
  "auto_updates": true,
```

The Homebrew team have decided that trying to ascertain actual version numbers is infeasible, so I wouldn't expect osquery to do any differently https://docs.brew.sh/FAQ#why-arent-some-apps-included-during-brew-upgrade so I feel like that rules out the original post's alternative number 2 (parsing version fields).

**Comment by zwass** [comment]
Closed by #8520.
